### PR TITLE
[incubator-kie-drools-5716] [new-parser] mismatched input 'if'/'do' (…

### PR DESCRIPTION
**Pull request by tkobayas**
…support named consequences)

**Issue**:

* https://github.com/apache/incubator-kie-drools/issues/5716

### Before PR in `drools-model-codegen`
```
[ERROR] Tests run: 2434, Failures: 167, Errors: 0, Skipped: 9
```

### After PR in `drools-model-codegen`
```
[ERROR] Tests run: 2434, Failures: 141, Errors: 0, Skipped: 9
```

Fixed all tests in `org.drools.model.codegen.execmodel.NamedConsequencesTest`

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request and downstream checks</b>  
  - Push a new commit to the PR. An empty commit would be enough.

- for a <b>full downstream build</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- for <b>Jenkins PR check only</b>
  - If you are an ASF committer for KIE podling, login to Jenkins (https://ci-builds.apache.org/job/KIE/job/drools/), go to the specific PR job, and click on `Build Now` button.
</details>

### CODE REVIEW

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/Antlr4ParserStringUtils.java [hunk 1]
```diff
@@ -66,13 +66,13 @@ public static String getTokenTextPreservingWhitespace(ParserRuleContext ctx, Tok
...
     /**
-     * Just remove leading "then"
+     * Extract name from "then[name]" of RHS_NAMED_CONSEQUENCE_THEN
      */
-    public static String trimThen(String rhs) {
-        if (rhs.startsWith("then")) {
-            return rhs.substring("then".length());
+    public static String extractNamedConsequenceName(String namedConsequenceThen) {
+        if (namedConsequenceThen.toLowerCase().startsWith("then[") && namedConsequenceThen.endsWith("]")) {
+            return namedConsequenceThen.substring("then[".length(), namedConsequenceThen.length() - 1);
```
Comment (tkobayas) [inline drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/Antlr4ParserStringUtils.java:73]: The String manipulation is not very nice, but if we split `RHS_NAMED_CONSEQUENCE_THEN` into finer-grained tokens, Lexer and Parser would get more complex while we want to retrieve RHS as simple text as possible.

**Comment by tkobayas** [comment]
@yurloc @gitgabrio @mariofusco Please review, thanks!

**Review by mariofusco — APPROVED** [review]

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLLexer.g4 [hunk 2]
```diff
@@ -172,6 +172,8 @@ RHS_STRING_LITERAL
     | ('\'' ( DrlEscapeSequence | ~('\\'|'\'') )* '\'') { setText( normalizeString( getText() ) ); }
     ;
 
+RHS_NAMED_CONSEQUENCE_THEN : 'then' LBRACK IDENTIFIER RBRACK ;
```
Comment (yurloc) [inline drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLLexer.g4]:
```suggestion
RHS_NAMED_CONSEQUENCE_THEN : DRL_THEN LBRACK IDENTIFIER RBRACK ;
```

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/DRLVisitorImpl.java [hunk 3]
```diff
@@ -292,8 +294,17 @@ public RuleDescr visitRuledef(DRLParser.RuledefContext ctx) {
...
+            // default consequence
             ruleDescr.setConsequenceLocation(ctx.rhs().getStart().getLine(), ctx.rhs().getStart().getCharPositionInLine()); // location of "then"
-            ruleDescr.setConsequence(trimThen(getTokenTextPreservingWhitespace(ctx.rhs(), tokenStream))); // RHS is just a text
+            ruleDescr.setConsequence(getTokenTextPreservingWhitespace(ctx.rhs().defaultConsequence(), tokenStream)); // RHS is just a text
+
+            // named consequences
+            ctx.rhs().namedConsequence()
+                     .forEach(namedConsequnceCtx -> {
+                         String name = extractNamedConsequenceName(namedConsequnceCtx.RHS_NAMED_CONSEQUENCE_THEN().getText());
+                         String body = getTokenTextPreservingWhitespace(namedConsequnceCtx.namedConsequenceBody(), tokenStream);
```
Comment (yurloc) [inline drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/DRLVisitorImpl.java]:
```suggestion
                     .forEach(namedConsequenceCtx -> {
                         String name = extractNamedConsequenceName(namedConsequenceCtx.RHS_NAMED_CONSEQUENCE_THEN().getText());
                         String body = getTokenTextPreservingWhitespace(namedConsequenceCtx.namedConsequenceBody(), tokenStream);
```

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4 [hunk 4]
```diff
@@ -488,9 +508,13 @@ lhsAccumulate : (DRL_ACCUMULATE|DRL_ACC) LPAREN lhsAndDef (COMMA|SEMI)
...
 
-rhs : DRL_THEN consequence ;
+rhs : DRL_THEN defaultConsequence namedConsequence* ;
+
+defaultConsequence : ( RHS_STRING_LITERAL | RHS_CHUNK )* ;
 
-consequence : ( RHS_STRING_LITERAL | RHS_CHUNK )* ;
+// THEN LEFT_SQUARE ID RIGHT_SQUARE chunk
+namedConsequence : RHS_NAMED_CONSEQUENCE_THEN namedConsequenceBody ;
+namedConsequenceBody : ( RHS_STRING_LITERAL | RHS_CHUNK )* ;
```
Comment (yurloc) [inline drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4]:
```suggestion
namedConsequence : RHS_NAMED_CONSEQUENCE_THEN defaultConsequence ;
```

`namedConsequenceBody` is identical to `defaultConsequence` so why not reuse it?

**Review by yurloc — APPROVED** [review]
A few suggestions, otherwise LGTM. Thanks!

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLLexer.g4 (hunk 2 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLLexer.g4]: @yurloc Thanks!

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/DRLVisitorImpl.java (hunk 3 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/DRLVisitorImpl.java]: @yurloc Thanks!

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4 (hunk 4 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRLParser.g4]: @yurloc Thank you. I applied the suggestion with a small name change.

**Comment by yurloc** [comment]
Overall fixing progress: There were 375 failed tests.

**Pull request merged**